src/pool/cli.c

```diff
@@ -1454,6 +1422,452 @@ dc_pool_query(tse_task_t *task)
+	uuid_copy(in->tmi_op.pi_uuid, pool->dp_pool);
+	uuid_copy(in->tmi_op.pi_hdl, pool->dp_pool_hdl);
+	in->tmi_map_version = map_version;
+
+	rc = map_bulk_create(ctx, &in->tmi_map_bulk, &b,
+			     pool_buf_nr(pool->dp_map_sz));
+	if (rc != 0) {
+		crt_req_decref(c);
+		return rc;
+	}
+
+	*rpc = c;
+	*map_buf = b;
+	return 0;
+}
+
+static void
+destroy_map_refresh_rpc(crt_rpc_t *rpc, struct pool_buf *map_buf)
+{
+	struct pool_tgt_query_map_in *in = crt_req_get(rpc);
+
+	map_bulk_destroy(in->tmi_map_bulk, map_buf);
+	crt_req_decref(rpc);
+}
+
+struct map_refresh_cb_arg {
+	crt_rpc_t	       *mrc_rpc;
+	struct pool_buf	       *mrc_map_buf;
+};
+
+static int
+map_refresh_cb(tse_task_t *task, void *varg)
+{
+	struct map_refresh_cb_arg      *cb_arg = varg;
+	struct map_refresh_arg	       *arg;
+	struct dc_pool		       *pool;
+	struct pool_tgt_query_map_in   *in = crt_req_get(cb_arg->mrc_rpc);
+	struct pool_tgt_query_map_out  *out = crt_reply_get(cb_arg->mrc_rpc);
+	struct pool_map		       *map;
+	bool				reinit = false;
+	int				rc = task->dt_result;
+
+	arg = tse_task_buf_embedded(task, sizeof(*arg));
+	pool = arg->mra_pool;
+
+	/*
+	 * If it turns out below that we do need to update the cached pool map,
+	 * then holding the lock while doing so will be okay, since we probably
+	 * do not want other threads to proceed with a known-stale pool anyway.
+	 * Otherwise, we will release the lock quickly.
+	 */
+	D_RWLOCK_WRLOCK(&pool->dp_map_lock);
+
+	D_DEBUG(DB_MD, DF_UUID": %p: crt: "DF_RC"\n", DP_UUID(pool->dp_pool),
+		task, DP_RC(rc));
+	if (daos_rpc_retryable_rc(rc)) {
+		reinit = true;
+		goto out;
+	} else if (rc != 0) {
+		goto out;
+	}
+
+	rc = out->tmo_op.po_rc;
+	if (rc == -DER_ALREADY) {
+		/*
+		 * The server side does not have a version we are interested
+		 * in. If the rank has a version < the highest known version,
+		 * it has a stale version itself, for which we need to try
+		 * another one. If the cached pool map version is known to be
+		 * stale, we also need to retry. Otherwise, we are done.
+		 */
+		D_DEBUG(DB_MD,
+			DF_UUID": %p: no requested version from rank %u: "
+			"requested=%u known=%u remote=%u\n",
+			DP_UUID(pool->dp_pool), task,
+			cb_arg->mrc_rpc->cr_ep.ep_rank, in->tmi_map_version,
+			pool->dp_map_version_known, out->tmo_op.po_map_version);
+		if (out->tmo_op.po_map_version < pool->dp_map_version_known ||
+		    map_known_stale(pool))
+			reinit = true;
+		rc = 0;
+		goto out;
+	} else if (rc == -DER_TRUNC) {
+		/*
+		 * cb_arg->mrc_map_buf is not large enough. Retry with the size
+		 * suggested by the server side.
+		 */
+		D_DEBUG(DB_MD, DF_UUID": %p: map buf < required %u\n",
+			DP_UUID(pool->dp_pool), task, out->tmo_map_buf_size);
+		pool->dp_map_sz = out->tmo_map_buf_size;
+		reinit = true;
+		goto out;
+	} else if (rc != 0) {
+		D_ERROR(DF_UUID": failed to fetch pool map: "DF_RC"\n",
+			DP_UUID(pool->dp_pool), DP_RC(rc));
+		goto out;
+	}
+
+	if (out->tmo_op.po_map_version < pool->dp_map_version_known) {
```

Comment:
So I guess here the client _would_ have been satisfied with the response (out->tmo_op.po_map_version compared to the in->tmi_map_version) if nothing had changed while the RPC was being handled? But in the meantime it seems the client has learned about a newer map version and saved it in pool->dp_map_version_known?

Reply:
Yes. This is how we effectively merge multiple concurrent pool map refreshes into one.